[Python] Unable to find kernel when calling from inside another kernel

### Required prerequisites

- [x] Consult the [security policy](https://github.com/NVIDIA/cuda-quantum/security/policy). If reporting a security vulnerability, do not report the bug using this form. Use the process described in the policy to report the issue.
- [x] Make sure you've read the [documentation](https://nvidia.github.io/cuda-quantum/latest). Your issue may be addressed there.
- [x] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to verify that this hasn't already been reported. +1 or comment there if it has.
- [ ] If possible, make a PR with a failing test to give us a starting point to work on!

### Describe the bug

This bug might be related to #2344. In that issue, CUDA-Q is able to find a kernel when it shouldn't. Here it is not able to find the kernel when it should.

### Steps to reproduce the bug

Take the following directory structure:
```
|-- foo/
|   |-- bar/
|   |   |-- zaz.py
|   |
|   |-- __init__.py
|
|-- main.py
```

Where `foo/bar/zaz.py` contains:
```python
import cudaq

@cudaq.kernel
def kernel():
    cudaq.qubit()
```

`foo/bar/__init__.py` contains:
```python
from .bar import zaz
```

`main.py` contains:
```python
from foo import bar
import cudaq

@cudaq.kernel
def test0():
    bar.zaz.kernel()

counts = cudaq.sample(test0)
print(counts)
```

Running the code `python main.py` outputs:
```
cudaq.kernel.ast_bridge.CompilerError: main.py:6: error: bar.zaz.kernel is not a valid quantum kernel to call.
	 (offending source -> bar.zaz.kernel())
```

Interestingly, using `foo.bar.zaz.kernel()` or only `kernel()` will make it work. (Likely because of #2344)

### Expected behavior

This kernel should correctly execute.

### Is this a regression? If it is, put the last known working version (or commit) here.

Not a regression

### Environment

- **CUDA Quantum version**: proto-0.8.0 (https://github.com/NVIDIA/cuda-quantum a9e2324f6504f02e47a87e1de82c8f31cbb0f46b)
- **Python version**: 3.12
- **Operating system**: ubuntu 24.04



### Suggestions

_No response_

--- a/python/tests/kernel/issue_2344/hidden_module/__init__.py
+++ b/python/tests/kernel/issue_2344/hidden_module/__init__.py
@@ -1,3 +1,11 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
 import cudaq
 
 

--- a/python/tests/kernel/issue_2344/hidden_module/__init__.py
+++ b/python/tests/kernel/issue_2344/hidden_module/__init__.py
@@ -1,0 +1,6 @@
+import cudaq
+
+
+@cudaq.kernel
+def spooky_kernel():
+    cudaq.qubit()

--- a/python/tests/kernel/issue_2344/test_issue_2344.py
+++ b/python/tests/kernel/issue_2344/test_issue_2344.py
@@ -27,6 +27,7 @@ def clear_registries():
 
 
 def test_unqualified_cross_module_kernel_call_raises():
+
     @cudaq.kernel
     def test0():
         spooky_kernel()  # not imported — should not resolve

--- a/python/tests/kernel/issue_2344/test_issue_2344.py
+++ b/python/tests/kernel/issue_2344/test_issue_2344.py
@@ -1,0 +1,35 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+# Regression test for issue #2344:
+# A kernel defined in another module (hidden_module.spooky_kernel) must NOT be
+# callable without its module qualifier. The call `spooky_kernel()` inside
+# test0 should raise a NameError because spooky_kernel is not in scope.
+#
+# Package layout (relative to this file):
+#   hidden_module/__init__.py    <- defines @cudaq.kernel def spooky_kernel()
+
+import cudaq
+import pytest
+
+import hidden_module  # noqa: F401 — imported for kernel registration side effect
+
+
+@pytest.fixture(autouse=True)
+def clear_registries():
+    yield
+    cudaq.__clearKernelRegistries()
+
+
+def test_unqualified_cross_module_kernel_call_raises():
+    @cudaq.kernel
+    def test0():
+        spooky_kernel()  # not imported — should not resolve
+
+    with pytest.raises((NameError, RuntimeError)):
+        cudaq.sample(test0)

--- a/python/tests/kernel/issue_2346/algo_lib/__init__.py
+++ b/python/tests/kernel/issue_2346/algo_lib/__init__.py
@@ -1,1 +1,9 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
 from .gates import qft_ops

--- a/python/tests/kernel/issue_2346/algo_lib/__init__.py
+++ b/python/tests/kernel/issue_2346/algo_lib/__init__.py
@@ -1,0 +1,1 @@
+from .gates import qft_ops

--- a/python/tests/kernel/issue_2346/algo_lib/gates/__init__.py
+++ b/python/tests/kernel/issue_2346/algo_lib/gates/__init__.py
@@ -1,0 +1,7 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #

--- a/python/tests/kernel/issue_2346/algo_lib/gates/qft_ops.py
+++ b/python/tests/kernel/issue_2346/algo_lib/gates/qft_ops.py
@@ -1,0 +1,6 @@
+import cudaq
+
+
+@cudaq.kernel
+def qft_kernel():
+    cudaq.qubit()

--- a/python/tests/kernel/issue_2346/algo_lib/gates/qft_ops.py
+++ b/python/tests/kernel/issue_2346/algo_lib/gates/qft_ops.py
@@ -1,3 +1,11 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
 import cudaq
 
 

--- a/python/tests/kernel/issue_2346/test_issue_2346.py
+++ b/python/tests/kernel/issue_2346/test_issue_2346.py
@@ -29,6 +29,7 @@ def clear_registries():
 
 
 def test_kernel_call_via_partial_module_path():
+
     @cudaq.kernel
     def test0():
         gates.qft_ops.qft_kernel()

--- a/python/tests/kernel/issue_2346/test_issue_2346.py
+++ b/python/tests/kernel/issue_2346/test_issue_2346.py
@@ -1,0 +1,37 @@
+# ============================================================================ #
+# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
+# All rights reserved.                                                         #
+#                                                                              #
+# This source code and the accompanying materials are made available under     #
+# the terms of the Apache License 2.0 which accompanies this distribution.     #
+# ============================================================================ #
+
+# Regression test for issue #2346:
+# A kernel called via an intermediate import path (gates.qft_ops.qft_kernel)
+# must resolve even though it is registered under the fully-qualified path
+# (algo_lib.gates.qft_ops.qft_kernel).
+#
+# Package layout (relative to this file):
+#   algo_lib/__init__.py              <- from .gates import qft_ops
+#   algo_lib/gates/__init__.py
+#   algo_lib/gates/qft_ops.py         <- defines @cudaq.kernel def qft_kernel()
+
+import cudaq
+import pytest
+
+from algo_lib import gates
+
+
+@pytest.fixture(autouse=True)
+def clear_registries():
+    yield
+    cudaq.__clearKernelRegistries()
+
+
+def test_kernel_call_via_partial_module_path():
+    @cudaq.kernel
+    def test0():
+        gates.qft_ops.qft_kernel()
+
+    counts = cudaq.sample(test0)
+    assert '0' in counts

